src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/GeneratedCodeContainer.cs

```diff
@@ -39,34 +95,59 @@ public TService GetService<TService>()
             return default(TService);
         }
 
-        public void SetOutput(SourceText source, RazorCodeDocument codeDocument)
+        public void SetOutput(DocumentSnapshot document)
         {
-            Source = source;
-            Output = codeDocument.GetCSharpDocument();
+            lock (_setOutputLock)
+            {
+                if (!document.TryGetTextVersion(out var version))
+                {
+                    return;
+                }
 
-            _textContainer.SetText(SourceText.From(Output.GeneratedCode));
+                var newerVersion = SourceVersion.GetNewerVersion(version);
+                if (newerVersion == SourceVersion)
+                {
+                    // Version from the provided document is the same or older as the current SourceVersion.
+                    return;
+                }
+
+                if (!document.TryGetText(out var source) ||
+                    !document.TryGetGeneratedOutput(out var codeDocument))
+                {
+                    return;
+                }
+
+                _source = source;
+                _sourceVersion = version;
+                _output = codeDocument.GetCSharpDocument();
+                _latestDocument = document;
+                _textContainer.SetText(SourceText.From(Output.GeneratedCode));
+            }
         }
 
         public Task<ImmutableArray<SpanMapResult>> MapSpansAsync(
-                Document document,
-                IEnumerable<TextSpan> spans,
-                CancellationToken cancellationToken)
+            Document document,
+            IEnumerable<TextSpan> spans,
+            CancellationToken cancellationToken)
         {
-            if (Output == null)
+            lock (_setOutputLock)
```

Comment:
You really don't want this lock held for the whole method. This should grab the output and then release the lock.

Reply:
👍